src/olympia/amo/models.py

```diff
@@ -238,29 +248,77 @@ def _send_changes(self, old_attr, new_attr_kw):
                 sender=self.__class__,
             )
 
-    def save(self, *args, **kw):
+    def save(self, *args, **kwargs):
         """
         Save changes to the model instance.
 
         If _signal=False is in `kw` the on_change() callbacks won't be called.
         """
-        signal = kw.pop('_signal', True)
-        result = super(OnChangeMixin, self).save(*args, **kw)
+        # When saving an existing instance, if the caller didn't specify
+        # an explicit update_fields and _dynamic_update_fields is absent or
+        # True, we attempt to find out which fields were changed and only
+        # save those. This allows for slightly better performance as we don't
+        # keep re-saving the same data over and over again, but also avoids
+        # overwriting data that has changed in the meantime.
+        # Fields with auto_now=True will be included all the time.
+        #
+        # Note that deferred fields will be included in the list of changed
+        # fields if they are loaded afterwards, even if their value does not
+        # change.
+        if (
+            self.pk
+            # Just having self.pk is not enough, we only really want to catch
+            # UPDATE calls and the caller could be doing Model(pk=1).save().
+            # Django save() implementation uses the special _state attribute
+            # for this.
+            and self._state.adding is False
+            and kwargs.get('update_fields') is None
+            and kwargs.pop('_dynamic_update_fields', True)
+        ):
+            fields = [f.attname for f in self._meta.concrete_fields]
+            initial_attrs = {
+                k: v for k, v in self._initial_attrs.items() if k in fields
+            }
+            kwargs['update_fields'] = list(
+                dict(
+                    {(k, self.__dict__[k]) for k in initial_attrs}
+                    - set(initial_attrs.items())
+                    # Never include primary key field - it might be set to None
+                    # initially in initial_attrs right after a call to create()
+                    # even though self.pk is set.
+                    - set(((self._meta.pk.name, self.pk),))
+                ).keys()
+            ) + [f.name for f in self._meta.fields if getattr(f, 'auto_now', False)]
```

Comment:
Although clever, imo this chunk of code (single statement!) is difficult to parse - it's tuples inside a set inside a dict inside a list.  Could it be rewritten slightly? (maybe bring the tuples + sets out into an `updated_values` variable or something?)

Reply:
Yeah, it grew bit by bit and spiraled out of control... I'll split it into variables

Reply:
b86edcaa0d7026a2fa82068cdf8fc14267ef8486